pinsDebug with more features, uses less RAM

Bob-the-Kuhn opened the pull request:
I've just uploaded a major change to pinsDebug.

The big change was creating an array in FLASH that contained every active pin definition. That reduced the RAM memory usage considerably but increased the FLASH usage. RAM usage dropped from 738 bytes to 68 bytes.  FLASH usage increased by 1.8k from 10.2k to 12.0k.

Creating the array requires going through the pin list twice. Rather than having two copies of it in the code I moved the list out to another file (pinsDebug_list.h) and then just did two #includes.

From the user’s view they’ll see the following changes:
1. Now reports all the names assigned to a pin
2. The port is now reported in addition to the pin number.
3. When PWM0A & PWM1C share a pin, both PWMs are reported
4. More PWM/Timer info is reported

One new item that may cause some concern is the usage of the LINE predefined preprocessor macro. It may not be available if the Arduino IDE goes to a different compiler.

----

Rats!! - I see that changes from my M3, M4, M5 work got dragged along with the pinsDebug changes.

The only changes are to the pinsDebug.h file and the addition of the pinsDebug_list.h file.

Roxy-3D commented:
RAM is much more precious than FLASH.  So I agree with your thinking.   But I'm wondering if we should make this so it can be turned off completely for people that have small AVR processors?

Bob-the-Kuhn commented:
pinsDebug is the reporting part of the M43 command.  

Turning M43 on/off is done via the **PINS_DEBUGGING** flag close to the end of Configuration_adv.h.

M43 without the pretty reporting would require some code changes to Marlin_main.cpp

Here's the memory usage of the various options.
M43 off
FLASH: 52,576 
RAM:    2,807 

M43 on with current pinsDebug
FLASH: 62,826
RAM:    3,545

M43 on with proposed pinsDebug
FLASH: 64,632
RAM:    2,875

M43 on without pretty reporting
FLASH: 53,240
RAM:    2,813

thinkyhead commented:
Seems to have some conflicts. Please resolve and I'll review this soon.

thinkyhead commented on Marlin/Conditionals_post.h:
```diff
@@ -514,7 +514,7 @@
   #define HAS_DIGIPOTSS (PIN_EXISTS(DIGIPOTSS))
   #define HAS_BUZZER (PIN_EXISTS(BEEPER) || ENABLED(LCD_USE_I2C_BUZZER))
-  #define HAS_CASE_LIGHT (PIN_EXISTS(CASE_LIGHT))
+  #define HAS_CASE_LIGHT (PIN_EXISTS(CASE_LIGHT) && ENABLED(CASE_LIGHT_ENABLE))
```
Please make the case light enable part of this PR into a separate PR.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -73,6 +73,9 @@
  * M1   - Same as M0
+ * M3   - Turn spindle on, set spindle speed, set rotation to clockwise
+ * M4   - Turn spindle on, set spindle speed, set rotation to counter-clockwise
+ * M5   - Turn spindle off
```
Please remove all spindle-related changes from this PR and make them in your `M3`/`M4`/`M5` branch instead. In general you need to isolate your features into their own branches and avoid this kind of cross-contamination.

thinkyhead commented:
This PR is a mess. Please isolate features and clean up the commits in all your current PRs. Git from the command-line, use of `git rebase -i`, and so on are essential to being a good gitizen.

Bob-the-Kuhn commented:
Just did a squash.

It's ready for review.

Bob-the-Kuhn commented:
Too many irons in the file & I ended up with slightly different versions of pinsDebug_Teensyduino.h on PR 5665 and PR 5668.

With this commit & squash they are now the same.

Bob-the-Kuhn commented:
This PR is ready for review.

thinkyhead commented:
Looking pretty awesome. I think we can cut down on PROGMEM usage a bit more by making some of the macros into function calls (i.e., augmenting them with function calls). I will point out the code I mean on the diff page in a moment…

There's a multi-thermocouple PR in the works that will add 3 more pins for testing. We'll have to add those pins to this PR once that one is ready.

thinkyhead commented on Marlin/pinsDebug.h:
```diff
@@ -724,75 +178,130 @@ static bool pwm_status(uint8_t pin) {
+#define CS_MAKE(N) (TCCR##N##B & (_BV(CS##N##0) | _BV(CS##N##1) | _BV(CS##N##2)) )
 #define WGM_MAKE3(N) (((TCCR##N##B & _BV(WGM##N##2)) >> 1) | (TCCR##N##A & (_BV(WGM##N##0) | _BV(WGM##N##1))))
 #define WGM_MAKE4(N) (WGM_MAKE3(N) | (TCCR##N##B & _BV(WGM##N##3)) >> 1)
 #define TIMER_PREFIX(T,L,N) do{ \
```
`TIMER_PREFIX` is one of the macros that can be made more economical by calling a function, passing only the parameters that differ. The reason this would shrink the binary is that the compiler doesn't combine identical PROGMEM strings into a single string in the DATA block, as it does with heap strings.

Bob-the-Kuhn replied:
How did you envision replacing all the name creation in this macro?

As best I can tell I'll need to create arrays of pointers to the registers something like this:
```
OCR_array[][A-C] PROGMEM = {

   #if defined(TCCR0A) && defined(COM0A1)
      #if defined (TIMER0A)
        {&OCR0A,&OCR0B,0},
      #else
        {0,&OCR0B,0},
      #endif
    #else  
        {0,0,0},
    #endif    

    #if defined(TCCR1A) && defined(COM1A1)
     #if defined(COM1C1) && defined (TIMER1C)
       {&OCR1A,&OCR1B,&OCR1C},
     #else
       {&OCR1A,&OCR1B,0},
     #endif 
    #else
        {0,0,0},   
    #endif

    #if defined(TCCR2A) && defined(COM2A1)
      {&OCR2A,&OCR2B,0},
    #endif

    #if defined(TCCR3A) && defined(COM3A1)
      {&OCR3A,&OCR3B,0},
     #if defined(COM3C1)
      {&OCR3A,&OCR3B,&OCR3C},
     #endif
    #endif

    #ifdef TCCR4A
      {&OCR4A,&OCR4B,&OCR4C},
    #endif

    #if defined(TCCR5A) && defined(COM5A1)
      {&OCR5A,&OCR5B,&OCR5C},
    #endif
  } 
```

thinkyhead reviewed (commented):
`TIMER_PREFIX` might be the only one…

Bob-the-Kuhn commented:
I've just uploaded the reduced PROGMEM size version of pinsDebug.h

The new code saves about 6K over the first version.  It now uses about 3.8k LESS than the pinsDebug.h currently in RCBugFix

The memory cost to enable PINS_DEBUGGING is now:
.   current RCBugFix
.   + 9.9K PROGMEM       + 680 bytes RAM
.
.  latest version
.       + 6.1K PROGMEM           + 66 bytes RAM
.
.  savings
.         3.8K  PROGMEM             614 bytes RAM

----

Also updated the pins list from today's RCBugFix

When the names of the 3 new pins are available I'll do an update and then a squash.

thinkyhead committed:
pinsDebug with more features, uses less RAM

I've just uploaded a major change to pinsDebug.
The big change was creating an array in FLASH that contained every
active pin definition. That reduced the RAM memory usage considerably
but increased the FLASH usage.
Creating the array requires going through the pin list twice. Rather
than having two copies of it in the code I moved the list out to another
file (pinsDebug_list.h) and then just did two #includes.
From the user’s view they’ll see the following changes:
1. Now reports all the names assigned to a pin
2. The port is now reported in addition to the pin number.
3. When PWM0A & PWM1C share a pin, both PWMs are reported
4. More PWM/Timer info is reported
One new item that may cause some concern is the usage of the LINE
predefined preprocessor macro. It may not be available if the Arduino
IDE goes to a different compiler.

Includes support for 1284 & 1286 families.

Memory usage changes when enabling PINS_DEBUGGING:
ATmega2560
FLASH
.           without   52576
.           with new  64592
.           with old  62826
.           new-out   12016
.           old-out   10250
.           new-old   1766
.
RAM
.           without   2807
.           with new  2875
.           with old  3545
.           new-out   68
.           old-out   738
.           new-old   -670

==================================================================

minor changes - mostly formatting

1) added newline to end of teensyduino file

2) changed flag name from TEENSYDUINO to TEENSYDUINO_IDE.  Got warnings
about redefining TEENSYDUINO

3) removed some trailing spaces

reduce PROGMEM size & update pin list

Reduced PROGMEM usage by
1) converting often used macro to a function
2) moved as much as possible into the function

This required creating two arrays of address pointers for the PWM
registers.

==================================================================

update with new M3, M4, M5 pin names

==================================================================

report I/O status for unused/unknown pins

Bob-the-Kuhn commented:
updated pin list with laser names & rebased it

thinkyhead commented:
If this can be rebased and resolved, then I will clean it up and merge it pretty soon.

Bob-the-Kuhn commented:
Just added I/O info for the unassigned/unknown pins
